### PR TITLE
SparK Implementation

### PR DESCRIPTION
**Pull request by johnsutor**
This is a WIP implementation of the SparK algorithm, as requested by #1462

### CODE REVIEW

**Comment by guarin** [comment]
Thanks a lot for the PR! The review might take a while due to the upcoming holiday season.

File: lightly/models/sparse/encoder.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    Basic class to store the mask for the sparse model.
+    """
+
+    def __init__(self):
+        self.mask: Union[Tensor, None] = None
+
+
+class SparseEncoder(nn.Module):
+    def __init__(
+        self,
```
Comment (guarin) [inline lightly/models/sparse/encoder.py:34]: Looking at the code I don't think we need this class (yet). It also mixes sparsity logic with hierarchical features which will make the class hard to maintain. Hierarchical features are already natively supported by TIMM or can be handled with a hook in torchvision, I don't think we should reimplement the logic here.

IMO the sparsity logic should be directly added when one implements the final model. Let's say we add `SparK` to the benchmarks in `benchmarks/imagenet/resnet50`. Then I would imagine that the implementation looks like this:
```
class SparK(LightningModule):
    def __init__(...):
        self.backbone = resnet50()
        self.sparse_mask = SparseMask()
        self.sparse_backbone =  dense_model_to_sparse(self.backbone, self.sparse_mask)
    
    def training_step(...):
        self.sparse_mask.mask = random_mask()
        features = self.sparse_backbone(x)
        ...
```

If hierarchical features are needed then this logic should be handled by the non-sparse backbone.

File: lightly/models/spark.py
Comment (guarin) [inline lightly/models/spark.py:1]: Could you move this file to `lightly/models/modules/spark.py`? The models in `lightly/models` like `lightly/models/simclr.py` etc. are deprecated and we don't want to add more models there.

File: lightly/models/spark.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+import torch
+import torch.nn as nn
+from lightly.models.sparse.encoder import SparseResnet
+from lightly.models.modules.sparse.spark import SparseBatchNorm2d, SparseSyncBatchNorm2d, SparseConvNeXtLayerNorm
+
+def is_pow2n(n: int) -> bool:
+    return n > 0 and (n & (n - 1)) == 0
+
+
+class UNetBlock(nn.Module):
```
Comment (guarin) [inline lightly/models/spark.py:17]:
```suggestion
class SparKUNetBlock(nn.Module):
```

File: lightly/models/spark.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        a UNet block with 2x up sampling
+        """
+        super().__init__()
+        self.up_sample = nn.ConvTranspose2d(
+            cin, cin, kernel_size=4, stride=2, padding=1, bias=True
+        )
+        self.conv = nn.Sequential(
+            nn.Conv2d(cin, cin, kernel_size=3, stride=1, padding=1, bias=False),
+            bn2d(cin),
+            nn.ReLU6(inplace=True),
```
Comment (guarin) [inline lightly/models/spark.py:29]:
```suggestion
            nn.ReLU(inplace=True),
```

Typo?

File: lightly/models/spark.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+            nn.Conv2d(cin, cout, kernel_size=3, stride=1, padding=1, bias=False),
+            bn2d(cout),
+        )
+
+    def forward(self, x):
+        x = self.up_sample(x)
+        return self.conv(x)
+
+
+class LightDecoder(nn.Module):
```
Comment (guarin) [inline lightly/models/spark.py:39]:
```suggestion
class SparKLightDecoder(nn.Module):
```

File: lightly/models/spark.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    def __init__(self, up_sample_ratio, width=768, sync_batch_norm=True):
+        super().__init__()
```
Comment (guarin) [inline lightly/models/spark.py:41]:
```suggestion
    def __init__(self, up_sample_ratio, width=768):
```

It shouldn't be necessary to add `sync_batch_norm` as argument as this can be handled outside the class. Torch has utility functions to automatically replace batch norms with sync batch norm.

File: lightly/models/spark.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        for i, d in enumerate(self.dec):
+            if i < len(to_dec) and to_dec[i] is not None:
+                x = x + to_dec[i]
+            x = self.dec[i](x)
+        return self.proj(x)
+
+    def extra_repr(self) -> str:
+        return f"width={self.width}"
+
+    def initialize(self):
```
Comment (guarin) [inline lightly/models/spark.py:68]:
```suggestion
    def init_weights(self):
```

Make it clear that this only initializes the weights.

File: lightly/models/spark.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            return recon_loss
+        
+        masked_bchw = inp_bchw * active_b1hw
+        rec_bchw = self.unpatchify(rec * var + mean)
+        rec_or_inp = torch.where(active_b1hw, inp_bchw, rec_bchw)
+        return inp_bchw, masked_bchw, rec_or_inp
+
+    
+    def patchify(self, bchw):
+        p = self.downsample_ratio
```
Comment (guarin) [inline lightly/models/spark.py:229]: Patchify and unpatchify already exist in `lightly/models/utils.py`

File: lightly/models/spark.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+                kernel_size = 1 if i <= 0 else 3
+                densify_proj = nn.Conv2d(e_width, d_width, kernel_size=kernel_size, stride=1, padding=kernel_size // 2, bias=True)
+                print(f'[SparK.__init__, densify {i+1}/{self.hierarchy}]: densify_proj(ksz={kernel_size}, #para={sum(x.numel() for x in densify_proj.parameters()) / 1e6:.2f}M)')
+            self.densify_projs.append(densify_proj)
+            d_width //= 2
+        
+        print(f'[SparK.__init__] dims of mask_tokens={tuple(p.numel() for p in self.mask_tokens)}')
+        
+    
+    def mask(self, B: int, device, generator=None):
```
Comment (guarin) [inline lightly/models/spark.py:161]: I am not sure how the masking works exactly but we already have a bunch of masking functions here https://github.com/lightly-ai/lightly/blob/356ae56fd8f6d6bbc548d4d189d9444f780a62ca/lightly/models/utils.py#L571-L838

If none of the existing functions are compatible then please add a new function there.

File: lightly/models/spark.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+                nn.init.kaiming_normal_(m.weight, mode="fan_out", nonlinearity="relu")
+                if m.bias is not None:
+                    nn.init.constant_(m.bias, 0.0)
+            elif isinstance(
+                m, (nn.LayerNorm, nn.BatchNorm1d, nn.BatchNorm2d, nn.SyncBatchNorm)
+            ):
+                nn.init.constant_(m.bias, 0)
+                nn.init.constant_(m.weight, 1.0)
+
+class SparK(nn.Module):
```
Comment (guarin) [inline lightly/models/spark.py:88]: We shouldn't add this class directly to the package. The goal of Lightly is to provide building blocks for models and not the full model implementation, we only add those in the benchmarks (e.g. `benchmarks/imagenet/resnet50/simclr.py`). So this class should eventually move to the benchmarks module. The different functionalities inside the class (masking, patchify, etc.) should move to `lightly/models/utils.py`.

File: lightly/models/sparse/encoder.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+from torch import List, Optional, Tensor
+
+from lightly.models.modules.sparse.spark import (
+    SparseAvgPooling,
+    SparseBatchNorm2d,
+    SparseConv2d,
+    SparseMaxPooling,
+    SparseSyncBatchNorm2d,
+    SparseConvNeXtBlock,
+    SparseConvNeXtLayerNorm
```
Comment (guarin) [inline lightly/models/sparse/encoder.py:15]: `SparseConvNeXtLayerNorm ` is imported but not used

File: lightly/models/modules/sparse/spark.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+                return super(SparseConvNeXtLayerNorm, self).forward(x)
+
+    def __repr__(self):
+        return (
+            super(SparseConvNeXtLayerNorm, self).__repr__()[:-1]
+            + f', ch={self.data_format.split("_")[-1]}, sp={self.sparse})'
+        )
+
+
+class SparseConvNeXtBlock(nn.Module):
```
Comment (guarin) [inline lightly/models/modules/sparse/spark.py:189]: What is the difference between this ConvNeXt implementation and the ones from torchvision and timm? Would it be possible to inherit from torchvisions `CNBlock` instead as it is used in `dense_model_to_sparse`? This way we wouldn't have to implement `DropPath` ourselves.

File: lightly/models/sparse/encoder.py
Comment (guarin) [inline lightly/models/sparse/encoder.py:1]: Let's move code in this file to `lightly/models/modules/spark_sparse.py`

File: lightly/models/modules/sparse/spark.py
Comment (guarin) [inline lightly/models/modules/sparse/spark.py:1]: Let's move the code in this file to `lighlty/models/modules/spark_sparse.py`

File: lightly/models/sparse/decoder.py
Comment (guarin) [inline lightly/models/sparse/decoder.py:1]: Let's not add this emtpy file

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks again for the PR! This will take a while to get merged as it contains many changes. I left a lot of comments/questions regarding the various parts. Let us know if you need help, we can also support you in making the changes or take over parts of the PR if that helps.

**Comment by johnsutor** [comment]
Thanks for all the comments! I'll make some changes later today and I'll let you know how much help I'll need given the scope of this project.

**Comment by guarin** [comment]
Sure, contributions are very welcome :)

**Comment by gabrielfruet** [comment]
@guarin Starting new PR, for now, a draft. Fixing some things you pointed out. 

I will first fix those that do not imply in great modifications, then i will test and see if its working properly, model training and etc, then i will refactor it to adhere to better engineering patterns. Is that okay to you?
